Sources/AsyncHTTPClient/ConnectionPool/HTTPConnectionPool+HTTP1State.swift

```diff
@@ -0,0 +1,694 @@
+                if let failedWaiter = connectionState.failed() {
+                    requestAction = .failRequest(failedWaiter.request, error, cancelWaiter: failedWaiter.requestID)
+                }
+
+                if let nextWaiter = self.waiters.popFirst() {
+                    assert(self.connections.count == self.maximumConcurrentConnections,
+                           "Why do we have waiters, if we could open more connections?")
+
+                    let eventLoop = nextWaiter.eventLoopRequirement ?? connectionState.eventLoop
+                    let newConnectionState = HTTP1ConnectionState(
+                        connectionID: self.idGenerator.next(),
+                        eventLoop: eventLoop,
+                        waiter: nextWaiter
+                    )
+                    self.connections[index] = newConnectionState
+                    return .init(requestAction, .createConnection(newConnectionState.connectionID, on: eventLoop))
+                }
+
+                self.connections.remove(at: index)
+                return .init(requestAction, .none)
+
+            case .shuttingDown(unclean: let unclean):
+                guard connectionState.failed() == nil else {
+                    preconditionFailure("Expected to remove the waiter when shutdown is issued")
+                }
+
+                self.connections.remove(at: index)
+                let isShutdown: StateMachine.ConnectionAction.IsShutdown
+                if self.connections.isEmpty {
+                    self.state = .shutDown
+                    isShutdown = .yes(unclean: unclean)
+                } else {
+                    isShutdown = .no
+                }
+
+                // the cleanupAction here is pretty lazy :)
+                return .init(.none, .cleanupConnection(close: [], cancel: [], isShutdown: isShutdown))
+
+            case .shutDown:
+                preconditionFailure("The pool is already shutdown all connections must already been torn down")
+            }
+        }
+
+        mutating func connectionTimeout(_ connectionID: Connection.ID) -> Action {
+            guard let index = self.connections.firstIndex(where: { $0.connectionID == connectionID }) else {
+                // because of a race this connection (connection close runs against trigger of timeout)
+                // was already removed from the state machine.
+                return .init(.none, .none)
+            }
+
+            assert(self.state == .running, "If we are shutting down, we must not have any idle connections")
+
+            var connectionState = self.connections[index]
+            guard connectionState.isAvailable else {
+                // connection is not available anymore, we may have just leased it for a request
+                return .init(.none, .none)
+            }
+
+            assert(self.waiters.isEmpty, "We have an idle connection, that times out, but waiters? Something is very wrong!")
```

Comment:
Remember that timers can fire during the same tick as other work, so this is absolutely possible. This should be checked for and ignored.